### ISSUE TITLE
[Bug] Deleting lockfile causes assertion to fail

### ISSUE THREAD

**Issue by andrew-boyarshin** [yarnpkg/berry#755]
- [ ] I'd be willing to implement a fix

Deleting a lockfile causes:
```
D:\dev\projectname>yarn
Internal Error: Assertion failed: The package should have been detected as part of the project
    at Function.find (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:36:574360)
    at async c.execute (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:58:22393)
    at async c.validateAndExecute (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:24:40434)
    at async c.run (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:36:533800)
    at async c.runExit (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:36:533934)
```

The missing lockfile was a valid scenario in Yarn v1, the lockfile was properly regenerated.

Creating empty `yarn.lock` causes different kind of issue.

```
➤ YN0000: ┌ Resolution step
➤ YN0032: │ sharp@npm:0.23.4: Implicit dependencies on node-gyp are discouraged
➤ YN0032: │ nan@npm:2.14.0: Implicit dependencies on node-gyp are discouraged
➤ YN0032: │ nan@npm:2.14.0: Implicit dependencies on node-gyp are discouraged
➤ YN0001: │ Error: graphql@^14.0.0 | ^15.0.0-rc.1 isn't supported by any available resolver
    at t.MultiResolver.getResolverByDescriptor (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:24:41628)
    at t.MultiResolver.bindDescriptor (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:24:41082)
    at t.OverrideResolver.bindDescriptor (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:36:862333)
    at C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:36:583324
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Promise.all (index 254)
    at async N.resolveEverything (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:36:582677)
    at async C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:36:599416
    at async l.startTimerPromise (C:\Users\Andrew\AppData\Roaming\npm\node_modules\yarn\bin\yarn.js:24:45528)
➤ YN0000: └ Completed in 13.87s
➤ YN0000: Failed with errors in 13.89s
```

Feels like NPM Semver resolver is malfunctioning on empty lockfiles. Lockfile is not updated/regenerated with what it succeeded to resolve before failure.

 - OS: Windows 10
 - Node version v13.6.0
 - Yarn version 2.0.0-rc.27

**Comment by arcanis** [comment]
Can you check whether you have a lockfile higher up in your filesystem hierarchy? For example in /home or /tmp, depending on your setup?

**Comment by andrew-boyarshin** [comment]
@arcanis Indeed I had (unexpectedly, to be honest). Unfortunately, I've already reverted to legacy Yarn due to #752, so checking this might take time (a couple of days).

**Comment by arcanis** [comment]
No worry - also note that `^14.0.0 | ^15.0.0-rc.1` is not a valid semver range. It should be `||`.

I'll close this issue but also look improving the first error message you got.

**Comment by nfarina** [comment]
Also encountered this due to a spurious `yarn.lock` in my home directory. Probably ended up there due to opening a new terminal window and typing `yarn` thinking I was in my project directory. Might be a common thing for users to have.

**Comment by adamwdennis** [comment]
I've moved back to `npm install` until this is resolved.

**Comment by hangtwenty** [comment]
> Also encountered this due to a spurious `yarn.lock` in my home directory. Probably ended up there due to opening a new terminal window and typing `yarn` thinking I was in my project directory. Might be a common thing for users to have.

Thank you for mention this @nfarina, I've spent hours trying to figure this out & you saved me from many more! None of the usual things I try were working.

In my case I had a `package.json` and `package-lock.json` in my home directory. I did a couple of other things ... `npm install -g yarn@1.22.x`, which didn't fully do the trick, but then `npx yarn install` seemed to? After that, yarn has been working great again. Itried reverting some steps and redoing them, but I couldn't reproduce the problem anymore. Still, maybe these steps could be of use to someone else troubleshooting.
